### PR TITLE
Prevent screen capture on Windows

### LINKED ISSUE #5859: Prevent Screenshotting of Sensitive Data [Security]
## Summary
[TIP]:  # ( DO NOT include screenshots of your actual database! )
[NOTE]: # ( Provide a brief overview of what the new feature is all about )

Currently, malware/spyware can just screen-capture sensitive data displayed by KeePassXC, including the master password or just the length of it (depending on eye button state).

Windows provides an API to prevent that, which should be used: [`SetWindowDisplayAffinity(hWnd, WDA_EXCLUDEFROMCAPTURE)`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowdisplayaffinity). This will just make the whole window black in screenshots.

To aid software documentation, articles about the software and such, it seems to be necessary to make the behavior adjustable.

There could be the following three modes of screenshottability:

- <q>Always</q> &mdash; As currently. When the user switched to this mode, which could have been temporarily necessary, just the next app start with this mode could trigger a dialog that asks the user whether they want to keep that mode. (Could be implemented by generically saving the last session's mode and comparing.)
- <q>Never</q> &mdash; If the user prefers, in case of a malware attack, to not even leak information about what their database contains. Should be the default in my opinion.
- <q>When Not Showing Passwords</q> &mdash; Makes an app window only unscreenshottable if it shows the most sensitive of data, which would include passwords (also in password generator; also key files, hardware keys or more?). To keep the window screenshottable as much as possible, it could just be made unscreenshottable when a password field contains more than zero characters. To prevent leaking the length of the all-determining master password, it could be argued that even password fields that just display dots instead of the actual characters should trigger unscreenshottability.
  - This mode could be left out, if it's not deemed important enough, because it would significantly increase implementation complexity.

If you decide against the last mode, a single menu entry may be the best: "Allow Screenshots Till Next Start", which will change to "Prevent Screenshots"/"Prevent Screenshots Again", if selected. (Behavior self-explanatory; not explained by modes above.)

## Examples
[NOTE]: # ( Show us a picture or mock-up of your proposal )

[How a window looks like when it's not screenshottable.](https://duckduckgo.com/?q=SetWindowDisplayAffinity&t=h_&iax=images&ia=images)

New menu entries for behavior adjustment:

```
Database   Entries   Groups   Tools   View   Help
                                      +-------------------+
                                      | [...]             |
                                      | Screenshottable > |+------------------------------+
                                      +-------------------+| • Never                      |
                                                           |   When Not Showing Passwords |
                                                           |   Always                     |
                                                           +------------------------------+
```

```
Database   Entries   Groups   Tools   View   Help
                                      +-----------------------------------+
                                      | [...]                             |
                                      | Allow Screenshots Till Next Start |
                                      +-----------------------------------+
```

```
Database   Entries   Groups   Tools   View   Help
                                      +---------------------+
                                      | [...]               |
                                      | Prevent Screenshots |
                                      +---------------------+
```

## Context
[NOTE]: # ( Why does this feature matter to you? What unique circumstances do you have? )

This is about one vector where protection of user data is overdue should be the default behavior.

### PR DESCRIPTION
**Pull request by smlu**
On Windows this PR makes KeePassXC main window and all it's child windows invisible for screen recording and screen shooting.
This is done by capturing native events `WM_NCCREATE` & `WM_ENTERIDLE`, and set `WDA_EXCLUDEFROMCAPTURE` display affinity flag via [SetWindowDisplayAffinity](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowdisplayaffinity) to the window in action or modal dialog. If for any reason the screen capturing of an app is desired i.e. demonstration, app should be run with `--allow-screencapture` argument. ~~When screen capturing for the app is enabled a persistent warning is displayed to the user, and title bar text of every displayed window is modified to contain text `[Screen Capture Allowed]`. The title bar text is also modified to contain the same text~~ In-app warning is displayed in case of an error occurs during the call to [SetWindowDisplayAffinity](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowdisplayaffinity) function.

fix #5859

## Screenshots
Demonstration of stealth mode vs running with `--allow-screencapture` arg:
img

## Testing strategy
The implementation was tested using `Snipping Tool` app, making print screen via keyboard shortcut and using [ScreenToGif](https://www.screentogif.com/) screen recording app.

## Type of change
- ✅ New feature (change that adds functionality)

### CODE REVIEW

**Comment by smlu** [comment]
Didn't have time to test this, but on the macOS the same could possible be achieved via undocumented API function `CGSSetWindowCaptureExcludeShape`. Here is an example code sample: https://github.com/heardrwt/RHAdditions/blob/69bed50e5f70304d3c3cf64e74e2005a8ec525eb/RHAdditions/NSWindow%2BRHPreventCaptureAdditions.m

File: src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -607,6 +607,34 @@ MainWindow::MainWindow()
...
+        connect(this, &MainWindow::windowTitleChanged, [this](const QString& title){
+            if (osUtils->allowScreenCapture()) {
+                static const auto postfix = tr(" [Screen Capture Allowed]", "MainWindow");
+                if (!title.contains(postfix)) {
+                    setWindowTitle(title + postfix);
+                }
+            }
+        });
+    }
+
```
Comment (droidmonkey) [inline src/gui/MainWindow.cpp:637]: Please remove all of this, it is not necessary and will be incredibly annoying while debugging or doing documentation work. The user has to very explicitly add a command line flag to enable screenshots, that in and of itself is enough barrier and warning.

File: src/gui/osutils/winutils/WinUtils.cpp [hunk 2]
```diff
@@ -18,13 +18,30 @@
...
+    if (not title.contains(postfix)) {
+        title.append(postfix);
+        if (SetWindowTextW(hwnd, (LPCWSTR)title.utf16())) {
+            return true;
+        }
+        qCritical().nospace() << "SetWindowTextW failed. error_code: 0x" << hex << GetLastError();
+    }
+    return false;
+}
+
```
Comment (droidmonkey) [inline src/gui/osutils/winutils/WinUtils.cpp:44]: Remove this

File: src/gui/osutils/winutils/WinUtils.cpp [hunk 3]
```diff
@@ -79,6 +96,41 @@ bool WinUtils::nativeEventFilter(const QByteArray& eventType, void* message, lon
     case WM_HOTKEY:
         triggerGlobalShortcut(msg->wParam);
         break;
+    case WM_NCCREATE:
+    case WM_ENTERIDLE: // modal dialog box i.e.: save file dialog
+        if (not allowScreenCapture()) {
```
Comment (droidmonkey) [inline src/gui/osutils/winutils/WinUtils.cpp:101]: We don't use verbs

File: src/gui/osutils/winutils/WinUtils.cpp [hunk 4]
```diff
@@ -79,6 +96,41 @@ bool WinUtils::nativeEventFilter(const QByteArray& eventType, void* message, lon
...
+            if (msg->message != WM_ENTERIDLE || DWORD(msg->wParam) == MSGF_DIALOGBOX) {
+                HWND hwnd = msg->message == WM_ENTERIDLE ? (HWND)msg->lParam : msg->hwnd;
+                if (not SetWindowDisplayAffinity(hwnd, WDA_EXCLUDEFROMCAPTURE)) {
+                    qCritical().nospace() << "Couldn't set WDA_EXCLUDEFROMCAPTURE display affinity flag to window. error_code: 0x"
+                                          << hex << GetLastError();
```
Comment (droidmonkey) [inline src/gui/osutils/winutils/WinUtils.cpp:106]: This is the only relevant change, and the only desired change. Remove everything else.

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 2 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp:44]: This code is part of error handling in the case `SetWindowDisplayAffinity` fails (user didn't specify `--allow-screencapture` in this case). It notifies the user via window title text that the window is no longer protected from screen capturing. We could also show error message box instead but I think it's more invasive, and to some extent less informative as you can forget that the screen protection is not enabled after you close the error message.

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 3 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp:101]: Should I rename to `isScreenCaptureAllowed`?

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/gui/osutils/winutils/WinUtils.cpp:44]: I'd prefer to give the in-app warning instead.

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/gui/osutils/winutils/WinUtils.cpp:101]: Sorry I meant we don't use `not` we use `!`

Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp:101]: Oh  ok, I'll refactor the code.

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 2 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp:44]: Ok, we could do via MessageWidget. In this case the issue can be modal popup dialogs (i.e. file save/open dialogs) as they could cover the warning and user might miss them. But yeah, in most cases it shouldn't come to this warning (I hope I'm not wrong :) so I think this solution is acceptable.

Comment (droidmonkey) [reply src/gui/osutils/winutils/WinUtils.cpp:44]: I am totally unconcerned since the current state of the application has no restrictions.

**Comment by c3X69puyujLWudBjwQEjmTUgEXr49cGa74EqHiL** [comment]
this would be good to have

File: src/gui/MainWindow.cpp (hunk 1 above)
Comment (smlu) [reply src/gui/MainWindow.cpp]: Done

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 2 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp]: The error handling is now resolved via in app warning message.

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 3 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp]: Done

File: src/gui/osutils/winutils/WinUtils.cpp (hunk 4 above)
Comment (smlu) [reply src/gui/osutils/winutils/WinUtils.cpp]: I cleaned the code, please verify if anything else should be changed.

**Comment by droidmonkey** [comment]
I greatly simplified your code and made it far less Windows centric.

**Review by droidmonkey — APPROVED** [review]

**Comment by droidmonkey** [comment]
Got it! macOS screenshot protection enabled :-D

**Comment by smlu** [comment]
Awesome!

**Comment by smlu** [comment]
> 
> 
> I greatly simplified your code and made it far less Windows centric. Will see if the macOS implementation can be added as well.

I've tested your changes and I found 1 problem. On windows, the modal windows of system file dialogs are not hidden from being captured by the screen recorder. For example when you browse for DB file or key file. I don't know if this is also the problem on macOS. This was the reason why I decided to handle prevention of app's screen being captured via native event filter. On Windows, I couldn't find any other way to do this through Qt framework and capture the native modal dialog belonging to the app being displayed. You don't get any signal or event whatsoever to notify you that another system dialog has opened that belongs to your app. So, the only solution I saw was to capture the native event [WM_ENTERIDLE](https://docs.microsoft.com/en-us/windows/win32/dlgbox/wm-enteridle) ([WinUtils.cpp#L105](https://github.com/smlu/keepassxc/blob/win-dev/src/gui/osutils/winutils/WinUtils.cpp#L105)) and hide the dialog that made the app idle. It's also possible there is some other solution to this problem, and I just didn't find it.

File: src/main.cpp [hunk 5]
```diff
@@ -132,6 +140,22 @@ int main(int argc, char** argv)
...
+        // This ensures any top-level windows (Main Window, Modal Dialogs, etc.) are excluded from screenshots
+        QObject::connect(&app, &QGuiApplication::focusWindowChanged, &mainWindow, [&](QWindow* window) {
+            if (window) {
+                if (!osUtils->setPreventScreenCapture(window, true)) {
+                    mainWindow.displayGlobalMessage(
+                        QObject::tr("Warning: Failed to prevent screenshots on a top level window!"),
+                        MessageWidget::Error);
+                }
+            }
+        });
```
Comment (smlu) [inline src/main.cpp:155]: This could be even more simplified by just hiding `MaingWindow` in each implementation (`getMainWindow()`). So you wouldn't need this connection being initialized here and `setPreventScreenCapture` can then just accept a single boolean parameter.
At least this is true on Windows, I haven't checked with macOS.

Comment (droidmonkey) [reply src/main.cpp:155]: I am unconcerned with file choice dialogs, those are actually owned by the OS and not KeePassXC process. We have to do more than mainwindow though to account for Auto-Type picker dialog, browser choice selection, totp dialogs, etc.

Comment (smlu) [reply src/main.cpp:155]: You're right about this connection. I've missed the fact that `SetWindowDisplayAffinity` sets affinity flag only for the targeted window and not also for the windows that might be associated with it. Unfortunately yes, Qt doesn't provide any signal to notify you about new window being created. I've also tried to override `childEvent` && `event` for `Application` & `MainWindow` classes and set affinity flag to new child windows & widgets, but it turns out not all windows are processed through these event functions. So I guess this connection here has to stay.

For the system dialogs not being hidden, I've also forgot to mentioned that the app's menu widgets and all context widgets are also not hidden. I don't know if you noticed this. I'm not saying it's crucial for the app security but this does leak some information, like the name of the database, location where it is stored (i.e. cloud storage), does the entry has TOTP set etc... 
For these reasons and the benefit of better confidentiality, I would rather then go with more complex implementation i.e. per platform window handling.

**Comment by thetuxinator** [comment]
Because of this feature, screensharing for example using Teams or Skype and co is no more possible as the Keeshare Window is also shown as a black window only in those cases so there should be a way to disable and enable it.

**Comment by thetuxinator** [comment]
Just discovered --allow-screencapture option from above, which also works in those cases

**Comment by 2BeeOrNot** [comment]
Just tried 2.7.0 on macOS and wanted to make a screen capture to report a display issue and had a what the heck moment. I can't make a screen capture. Poking around and searching here I found this PR. Can/should there be an option in KPXC to allow screen captures instead of using terminal to start KPXC with --allow-screencapture? Won't this confuse users who want to make screen captures for training, documentation, whatever? How will users know to use --allow-screencapture from terminal?

**Comment by droidmonkey** [comment]
I just added it to our documentation so that's how.

**Comment by Kellett** [comment]
Just upgraded to 2.7 and had to spend a while trying to work out why this wasn't visible on any remote machines. Is there a setting to always open with "--allow-screencapture" without having to edit the auto-start entries / shortcuts?

**Comment by phoerious** [comment]
You add the flag to KeePassXC's auto-start entry.

**Comment by michaelk83** [comment]
> You add the flag to KeePassXC's auto-start entry.

Or preferably, make a separate shortcut for only when that switch is actually needed.

**Comment by droidmonkey** [comment]
No you need to edit the startup entries

**Comment by droidmonkey** [comment]
Create your own shortcuts, we aren't in the business of making it easy to circumvent a security feature. Actually wanting to screen record your database is very rare.

**Comment by Kellett** [comment]
I don't want to screen record the database. That probably is a rare thing. I just want to use it over a remote desktop connection. I can edit the shortcuts if that is the only option. I was just hoping there was a easier option.

**Comment by qohjjang** [comment]
> Create your own shortcuts, we aren't in the business of making it easy to circumvent a security feature. Actually wanting to screen record your database is very rare.

Thanks for working hard to make this feature available.
It is actually very frustrating when you want to capture a screenshot of an entity (or remote access for that matter). Adding a shortcut to open it again is also not very good solution, as you may forget to close it again to prevent it.
As a user perspective, there should be an option that a user can tick it on/off on Setting or menu somewhere.

**Comment by marketbusiness** [comment]
> Just discovered --allow-screencapture option from above, which also works in those cases

Hello, how do you use the --allow-screencapture command on a windows shortcut. I use keepass via team wiever but with this option set up it is impossible for me to use it

**Comment by michaelk83** [comment]
tldr: https://superuser.com/a/29571/1271887
step by step: https://www.digitalcitizen.life/shortcut-arguments-parameters-windows/
